lib/oplog/tail.go

```diff
@@ -43,19 +43,30 @@ type rawOplogEntryID struct {
 
 const requeryDuration = time.Second
 
-var metricOplogEntriesReceived = promauto.NewCounterVec(prometheus.CounterOpts{
-	Namespace: "otr",
-	Subsystem: "oplog",
-	Name:      "entries_received",
-	Help:      "Oplog entries received, partitioned by database and status",
-}, []string{"database", "status"})
-
-var metricOplogEntriesReceivedSize = promauto.NewCounterVec(prometheus.CounterOpts{
-	Namespace: "otr",
-	Subsystem: "oplog",
-	Name:      "entries_received_size",
-	Help:      "Size of oplog entries received in bytes, partitioned by database",
-}, []string{"database"})
+var (
+	// Deprecated: use metricOplogEntriesBySize instead
+	metricOplogEntriesReceived = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "otr",
+		Subsystem: "oplog",
+		Name:      "entries_received",
+		Help:      "Oplog entries received, partitioned by database and status",
+	}, []string{"database", "status"})
+
+	// Deprecated: use metricOplogEntriesBySize instead
+	metricOplogEntriesReceivedSize = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "otr",
+		Subsystem: "oplog",
+		Name:      "entries_received_size",
+		Help:      "Size of oplog entries received in bytes, partitioned by database",
+	}, []string{"database"})
+
+	metricOplogEntriesBySize = promauto.NewHistogramVec(prometheus.HistogramOpts{
```

Comment:
I think moving the existing two metrics to a histogram is good, but I'm not sure it entirely does what we want in terms of being able to monitor the maximum size of document sent through oplogtoredis. Also, we'll want to use custom histogram buckets, the [default ones](https://godoc.org/github.com/prometheus/client_golang/prometheus#pkg-variables) are designed for response times in seconds rather than document size in bytes.

I'm not sure Prometheus supports a maximum metric out of the box; let's talk on Tuesday about how we can handle that.

Reply:
I guess we can't do maximum over an arbitrary window, but we could compute a gauge metric that reports the maximum size of document observed over the previous 1-minute bucket -- so we'd take the current timestamp, truncate to 1-minute, and then report as a gauge the maximum value from the previous minute-long bucket

Reply:
This is now implemented with `IntervalMaxMetric(Vec)`, a custom Prometheus `Collector` that supports this behavior

Reply:
And I went with exponential request size buckets (up to 2^32)